jme3-effects/src/main/java/com/jme3/post/filters/SsrFilter.java

```diff
@@ -0,0 +1,305 @@
+    protected void initFilter(AssetManager manager, RenderManager renderManager, ViewPort vp, int w, int h) {
+        
+        this.renderManager = renderManager;
+        this.viewPort = vp;
+        int screenWidth = w;
+        int screenHeight = h;
+                postRenderPasses = new ArrayList<>();
+
+
+        frustumNearFar = new Vector2f();
+
+        float farY = (vp.getCamera().getFrustumTop() / vp.getCamera().getFrustumNear()) * vp.getCamera().getFrustumFar();
+        float farX = farY * ((float) screenWidth / (float) screenHeight);
+        frustumNearFar.x = vp.getCamera().getFrustumNear();
+        frustumNearFar.y = vp.getCamera().getFrustumFar();
+        
+        
+//        if(!approximateNormals){
+            normalPass = new Pass();
+            normalPass.init(renderManager.getRenderer(), (int) (screenWidth / downSampleFactor), (int) (screenHeight / downSampleFactor), Image.Format.RGBA8, Image.Format.Depth);
+//        }
+
+        ssrMaterial = new Material(manager, "Common/MatDefs/SSR/ssr.j3md"); 
+//        if(!approximateNormals){
+            ssrMaterial.setTexture("Normals", normalPass.getRenderedTexture());
+//        }
+        ssrMaterial.setInt("RaySamples", raySteps);
+        ssrMaterial.setInt("NearbySamples", sampleNearby ? 4 : 0);
+        ssrMaterial.setFloat("StepLength", stepLength);
+        ssrMaterial.setFloat("ReflectionFactor", reflectionFactor);
+        ssrMaterial.setBoolean("ApproximateNormals", approximateNormals);
+        ssrMaterial.setVector2("NearReflectionsFade", nearFade);
+        ssrMaterial.setVector2("FarReflectionsFade", farFade);
+        
+        ssrPass = new Pass("SSR pass") {
+
+            @Override
+            public boolean requiresDepthAsTexture() {
+                return true;
+            }
+
+            @Override
+            public boolean requiresSceneAsTexture() {
+                return true; //To change body of generated methods, choose Tools | Templates.
+            }
+            
+            
+        };
+
+        ssrPass.init(renderManager.getRenderer(), (int) (screenWidth / downSampleFactor), (int) (screenHeight / downSampleFactor), Image.Format.RGBA8, Image.Format.Depth, 1, ssrMaterial);
+        postRenderPasses.add(ssrPass);
+
+        material = new Material(manager, "Common/MatDefs/SSR/ssrBlur.j3md"); 
+        material.setTexture("SSR", ssrPass.getRenderedTexture());
+        material.setBoolean("Horizontal", true);
+        material.setBoolean("FastBlur", fastBlur);
+        material.setFloat("BlurScale", blurScale);
+        material.setFloat("Sigma", sigma);
+        
+}
+    
+    @Override
+    protected void postQueue(RenderQueue queue) {
+        if(!approximateNormals) {
+            Renderer r = renderManager.getRenderer();
+            r.setFrameBuffer(normalPass.getRenderFrameBuffer());
+            renderManager.getRenderer().clearBuffers(true, true, true);
+            renderManager.setForcedTechnique("PreNormalPass");
```

Comment:
Those normals are in view space, the shader wants world space normals.

Reply:
I know. With the changes to the shader I can't see a difference between approximateNormals and the normalPass.

Reply:
Well, with this code, if you disable approximated normals, it will look very wrong. But the idea is that with the approximated normals you can only get an approximation of the surface normals, while with the normal pass you can get the real world space normals with the normal map applied. 
The shader can also reconstruct the normal Z and you can use the free Z channel of the normal map to store the glossiness (eg for PBR materials).

But the PreNormalPass doesn't do any of that, so that pass should be replaced with something else. Or we could decide to disable the normal pass entirely and allow only approximated normals, this could be a valid or maybe also preferable solution, since rerendering the scene only to extract the normals is not a good idea in any case, this should be done with MRT instead but here i think we are hitting the limitations of jme post processing.

What do you think?